# Data Cleaning in Pandas

- The data that we will be cleaning is a Customer Call List dataframe, imported from an Excel file.
- The dataframe will be used to provide a contact list of customers to call for a contact centre and thus will need to be cleaned to only provide them with relevant information specific to the needs of a call centre.

- First, we import pandas 

In [361]:
import pandas as pd

- We can use pd.read_excel(r"filelocation") to read in the data.

In [362]:
df = pd.read_excel(r"C:\Users\Alicja\Desktop\Charles\data_analytics\da_courses_and_projects\python\courses\pandas\data_cleaning_pandas\raw_data\Customer Call List.xlsx")
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   
20        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   
20            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True  
20               True

- **There are many issues with the data**:
    1. Duplicate rows (rows 19 and 20 are identical) will need to be dropped.
    2. The last column (`'Not_Useful_Column'`) is not useful and thus will need to be dropped.
    3. Non-alphanumeric characters in the `'Last_Name'` column will need to be stripped.
    4. Non-numeric character in the `'Phone_Number'` column will need to be removed and the numbers standardised.
    5. Some data in the `'Address'` column contains zipcodes and the States/Commonwealth, whereas others are shorter, this will need to be stanbdardised and maybe columns split to enable a column for zipcodes and states or at the very least just to standardise the `'Address'` column.
    6. In the `'Paying Customer'` column, some are `'Yes/No'`, while others are `'Y/N'` this will need to be standardised. The column name will also need correcting to include an underscore for standardisation.
    7. `'NaN'` and `'N/a'` (`Null`) values within the dataframe will need to be removed.
    8. Some values within the `'Do_Not_Contact'` column are `'Yes'` or `'Y'` and thus, these rows will need to be removed, along with any that contain no value within the `'Phone_Number'` column.

## 1. Drop duplicates:

In [363]:
df.drop_duplicates()

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

- Using df.drop_duplicates() removed the duplicate row.
- As it worked, we will assign the result to the variable df.

In [364]:
df = df.drop_duplicates()
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

## 2. Drop Column

- We need to remove the last column (`'Not_Useful_Column'`) and to do so, we will use `df.drop(columns = 'Column_Name')`

In [365]:
df.drop(columns = 'Not_Useful_Column')

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

- That successfully removed the `'Not_Useful_Column'` column and thus we will assign this to `df` to update the dataframe.

In [366]:
df = df.drop(columns = 'Not_Useful_Column')
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

## 3. Strip Non-Alphanumeric Characters from 'Last_Name'

- **Important notes:**
  1. When using `strip()`, you **MUST** specify the column name(s), or it will apply to all columns, potentially causing significant issues.
  2. `strip()` only targets characters at the start and end of strings, making it unsuitable for removing unwanted characters from the middle; for middle characters, `replace()` can be used instead.
- **`df[['Last_Name', 'First_Name']]`**: Selects multiple columns (e.g., "Last_Name" and "First_Name") as a pandas DataFrame.
- **`.str`**: Accessor to apply string methods to each element across the selected columns.
- **`.strip()`**: Removes leading and trailing specified non-alphanumeric characters (e.g., `df[['Last_Name', 'First_Name']].str.strip('./_,+')` removes '.' and '/' from each string's ends); note that `lstrip()` and `rstrip()` can remove characters from the left or right only, but we will use `strip()` for both ends.
- **Combined**: Cleans specified non-alphanumeric characters from the start and end of all values in the selected column(s); reassignable to update the columns (e.g., `df['Last_Name'] = df['Last_Name'].str.strip('123./_,+')`).

In [367]:
df['Last_Name'].str.strip('./_')

0        Baggins
1          Nadir
2          White
3        Schrute
4           Snow
5        Swanson
6         Winger
7         Holmes
8            NaN
9         Parker
10        Gamgee
11        Potter
12        Draper
13         Knope
14    Flenderson
15       Weasley
16         Scott
17          Kent
18        Braton
19     Skywalker
Name: Last_Name, dtype: object

- **Specifying the unwanted characters together** e.g. **`strip('./_')`** can be performed to save iterations.
- We can then assign the output to the column in order to update the cleaned column, using `df['Last_Name'] = df['Last_Name'].str.strip('./_')`

In [368]:
df['Last_Name'] = df['Last_Name'].str.strip('./_')
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger           NaN   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf         NaN           N/a   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee           NaN   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton           N/a   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

## 4. Cleaning and Standardising Phone Numbers

In [369]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger           NaN   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf         NaN           N/a   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee           NaN   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton           N/a   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

- The format that we will be standardising phone numbers to: 012-345-6789.
- We will remove all non-numeric values and strip it down to just the numbers to begin with.
- After that, we will format the entire column with the same format.

In [370]:
df['Phone_Number'].str.replace('[^a-zA-Z0-9]','', regex = True)

0     1235455421
1     1236439775
2            NaN
3     1235432345
4     8766783469
5     3047622467
6            NaN
7     8766783469
8             Na
9     1235455421
10           NaN
11           NaN
12    1235432345
13    8766783469
14    3047622467
15    1235455421
16    1236439775
17           NaN
18            Na
19    8766783469
Name: Phone_Number, dtype: object

- That worked to remove all non-alphanumeric characters from each string value within the `'Phone_Number`'` column.
- We will now assign it to the column:

In [371]:
df['Phone_Number'] = df['Phone_Number'].str.replace('[^a-zA-Z0-9]','', regex = True)
df

CustomerID First_Name   Last_Name Phone_Number  \
0         1001      Frodo     Baggins   1235455421   
1         1002       Abed       Nadir   1236439775   
2         1003     Walter       White          NaN   
3         1004     Dwight     Schrute   1235432345   
4         1005        Jon        Snow   8766783469   
5         1006        Ron     Swanson   3047622467   
6         1007       Jeff      Winger          NaN   
7         1008   Sherlock      Holmes   8766783469   
8         1009    Gandalf         NaN           Na   
9         1010      Peter      Parker   1235455421   
10        1011    Samwise      Gamgee          NaN   
11        1012      Harry      Potter          NaN   
12        1013        Don      Draper   1235432345   
13        1014     Leslie       Knope   8766783469   
14        1015       Toby  Flenderson   3047622467   
15        1016        Ron     Weasley   1235455421   
16        1017   Michael        Scott   1236439775   
17        1018      Clark        Kent          NaN   
18        1019      Creed      Braton           Na   
19        1020     Anakin   Skywalker   8766783469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

- Now we will format it to (012-345-6789) using a lambda.

In [372]:
df['Phone_Number'].apply(lambda x: x[0:3] + '-' + x[3:6] + '-' + x[6:10])

TypeError: 'float' object is not subscriptable

- This outputs an error because the float is not subscriptable, thus, we need to convert that data to string first, using str(df['Phone_Number']) does not work.
- Instead, we can use apply(lambda x: str(x)) and then proceed to format them with hyphens at the correct intervals.

In [373]:
df['Phone_Number'] = df['Phone_Number'].apply(lambda x: str(x))

- Now the data within the column has been converted to strings, we can proceed to formatting.

In [374]:
df['Phone_Number'].apply(lambda x: x[0:3] + '-' + x[3:6] + '-' + x[6:10])

0     123-545-5421
1     123-643-9775
2            nan--
3     123-543-2345
4     876-678-3469
5     304-762-2467
6            nan--
7     876-678-3469
8             Na--
9     123-545-5421
10           nan--
11           nan--
12    123-543-2345
13    876-678-3469
14    304-762-2467
15    123-545-5421
16    123-643-9775
17           nan--
18            Na--
19    876-678-3469
Name: Phone_Number, dtype: object

- That correctly formatted the `'Phone_Number'` column values, so we will assing it to the column.

In [375]:
df['Phone_Number'] = df['Phone_Number'].apply(lambda x: x[0:3] + '-' + x[3:6] + '-' + x[6:10])
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White         nan--   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger         nan--   
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN          Na--   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee         nan--   
11        1012      Harry      Potter         nan--   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent         nan--   
18        1019      Creed      Braton          Na--   
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

- We will now remove the `'nan--'` and `'Na--'` values.
- Starting with `'nan--'`

In [376]:
df['Phone_Number'].str.replace('nan--', '')

0     123-545-5421
1     123-643-9775
2                 
3     123-543-2345
4     876-678-3469
5     304-762-2467
6                 
7     876-678-3469
8             Na--
9     123-545-5421
10                
11                
12    123-543-2345
13    876-678-3469
14    304-762-2467
15    123-545-5421
16    123-643-9775
17                
18            Na--
19    876-678-3469
Name: Phone_Number, dtype: object

- That worked, we will assign it and repeat but for `'Na--'`

In [377]:
df['Phone_Number'] = df['Phone_Number'].str.replace('nan--', '')
df['Phone_Number'] = df['Phone_Number'].str.replace('Na--', '')
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

## 5. i) Splitting the `'Address'` Column

- We will use the split function and will split using commas as the delimiter.

In [378]:
df['Address'].str.split(',',n=2, expand=True)

0              1       2
0         123 Shire Lane          Shire    None
1    93 West Main Street           None    None
2     298 Drugs Driveway           None    None
3       980 Paper Avenue   Pennsylvania   18503
4       123 Dragons Road           None    None
5       768 City Parkway           None    None
6      1209 South Street           None    None
7          98 Clue Drive           None    None
8       123 Middle Earth           None    None
9       25th Main Street       New York    None
10        612 Shire Lane          Shire    None
11  2394 Hogwarts Avenue           None    None
12      2039 Main Street           None    None
13      343 City Parkway           None    None
14         214 HR Avenue           None    None
15  2395 Hogwarts Avenue           None    None
16      121 Paper Avenue   Pennsylvania    None
17       3498 Super Lane           None    None
18                   N/a           None    None
19     910 Tatooine Road       Tatooine    None

- That worked but it output them with indexes across the top as column names and we don't want these.
- We can ensure that when we assign the new data, we do not include those numerical headers/indexes by inputting the new column names as a list, i.e. instead of `df['Address']`, we would use `df[['Street_Address','State','Zip_Code']]`, this will create 3 new columns and retain the original `'Address'` column, which we will need to go back and remove afterwards.

In [379]:
df[['Street_Address', 'State', 'Zip_Code']] = df['Address'].str.split(',',n=2, expand=True)
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire     None  
1    93 West Main Street           None     None  
2     298 Drugs Driveway           None     None  
3       980 Paper Avenue   Pennsylvania    18503  
4       123 Dragons Road           None     None  
5       768 City Parkway           None     None  
6      1209 South Street           None     None  
7          98 Clue Drive           None     None  
8       123 Middle Earth           None     None  
9       25th Main Street       New York     None  
10        612 Shire Lane          Shire     None  
11  2394 Hogwarts Avenue           None     None  
12      2039 Main Street           None     None  
13      343 City Parkway           None     None  
14         214 HR Avenue           None     None  
15  2395 Hogwarts Avenue           None     None  
16      121 Paper Avenue   Pennsylvania     None  
17       3498 Super Lane           None     None  
18                   N/a           None     None  
19     910 Tatooine Road       Tatooine     None

## 5. ii) Dropping the Original `'Address Column'`

In [380]:
df.drop(columns = 'Address')

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421             Yes   
1         1002       Abed       Nadir  123-643-9775              No   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345             Yes   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467             Yes   
6         1007       Jeff      Winger                            No   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN                           Yes   
9         1010      Peter      Parker  123-545-5421             Yes   
10        1011    Samwise      Gamgee                           Yes   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345             Yes   
13        1014     Leslie       Knope  876-678-3469             Yes   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421              No   
16        1017   Michael        Scott  123-643-9775             Yes   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                           N/a   
19        1020     Anakin   Skywalker  876-678-3469             Yes   

   Do_Not_Contact        Street_Address          State Zip_Code  
0              No        123 Shire Lane          Shire     None  
1             Yes   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4              No      123 Dragons Road           None     None  
5             Yes      768 City Parkway           None     None  
6              No     1209 South Street           None     None  
7              No         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9              No      25th Main Street       New York     None  
10             No        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13             No      343 City Parkway           None     None  
14             No         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16             No      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18            Yes                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

- That worked, we can now assign that to the dataframe.

In [381]:
df = df.drop(columns = 'Address')
df

CustomerID First_Name   Last_Name  Phone_Number Paying Customer  \
0         1001      Frodo     Baggins  123-545-5421             Yes   
1         1002       Abed       Nadir  123-643-9775              No   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345             Yes   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467             Yes   
6         1007       Jeff      Winger                            No   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN                           Yes   
9         1010      Peter      Parker  123-545-5421             Yes   
10        1011    Samwise      Gamgee                           Yes   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345             Yes   
13        1014     Leslie       Knope  876-678-3469             Yes   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421              No   
16        1017   Michael        Scott  123-643-9775             Yes   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                           N/a   
19        1020     Anakin   Skywalker  876-678-3469             Yes   

   Do_Not_Contact        Street_Address          State Zip_Code  
0              No        123 Shire Lane          Shire     None  
1             Yes   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4              No      123 Dragons Road           None     None  
5             Yes      768 City Parkway           None     None  
6              No     1209 South Street           None     None  
7              No         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9              No      25th Main Street       New York     None  
10             No        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13             No      343 City Parkway           None     None  
14             No         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16             No      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18            Yes                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

## 6. i) Adding an undercore to the column name for 'Paying Customer'

- Due to the `'Paying Customer'` column containing a space and not an underscore like the others, we will rename it to standardise it.

In [382]:
df.rename(columns={'Paying Customer': 'Paying_Customer'})

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421             Yes   
1         1002       Abed       Nadir  123-643-9775              No   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345             Yes   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467             Yes   
6         1007       Jeff      Winger                            No   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN                           Yes   
9         1010      Peter      Parker  123-545-5421             Yes   
10        1011    Samwise      Gamgee                           Yes   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345             Yes   
13        1014     Leslie       Knope  876-678-3469             Yes   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421              No   
16        1017   Michael        Scott  123-643-9775             Yes   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                           N/a   
19        1020     Anakin   Skywalker  876-678-3469             Yes   

   Do_Not_Contact        Street_Address          State Zip_Code  
0              No        123 Shire Lane          Shire     None  
1             Yes   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4              No      123 Dragons Road           None     None  
5             Yes      768 City Parkway           None     None  
6              No     1209 South Street           None     None  
7              No         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9              No      25th Main Street       New York     None  
10             No        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13             No      343 City Parkway           None     None  
14             No         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16             No      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18            Yes                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

- That worked, we can then assign that change to the dataframe.

In [383]:
df = df.rename(columns={'Paying Customer': 'Paying_Customer'})
df

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421             Yes   
1         1002       Abed       Nadir  123-643-9775              No   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345             Yes   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467             Yes   
6         1007       Jeff      Winger                            No   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN                           Yes   
9         1010      Peter      Parker  123-545-5421             Yes   
10        1011    Samwise      Gamgee                           Yes   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345             Yes   
13        1014     Leslie       Knope  876-678-3469             Yes   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421              No   
16        1017   Michael        Scott  123-643-9775             Yes   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                           N/a   
19        1020     Anakin   Skywalker  876-678-3469             Yes   

   Do_Not_Contact        Street_Address          State Zip_Code  
0              No        123 Shire Lane          Shire     None  
1             Yes   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4              No      123 Dragons Road           None     None  
5             Yes      768 City Parkway           None     None  
6              No     1209 South Street           None     None  
7              No         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9              No      25th Main Street       New York     None  
10             No        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13             No      343 City Parkway           None     None  
14             No         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16             No      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18            Yes                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

## 6. ii) Using `df.str.replace()` to convert 'Yes' to 'Y' and 'No' to 'N' values within 'Paying_Customer' and 'Do_Not_Contact'

In [384]:
df['Paying_Customer'].str.replace('Yes','Y').str.replace('No','N')

0       Y
1       N
2       N
3       Y
4       Y
5       Y
6       N
7       N
8       Y
9       Y
10      Y
11      Y
12      Y
13      Y
14      N
15      N
16      Y
17      Y
18    N/a
19      Y
Name: Paying_Customer, dtype: object

- That worked, we can assign that to the column now and then repeat for the 'Do_Not_Contact' column.

In [385]:
df['Paying_Customer'] = df['Paying_Customer'].str.replace('Yes','Y').str.replace('No','N')
df

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1002       Abed       Nadir  123-643-9775               N   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345               Y   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467               Y   
6         1007       Jeff      Winger                             N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN                             Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee                             Y   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                           N/a   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0              No        123 Shire Lane          Shire     None  
1             Yes   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4              No      123 Dragons Road           None     None  
5             Yes      768 City Parkway           None     None  
6              No     1209 South Street           None     None  
7              No         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9              No      25th Main Street       New York     None  
10             No        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13             No      343 City Parkway           None     None  
14             No         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16             No      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18            Yes                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

- That worked, so we will repeat that full process for the 'Do_Not_Disturb' column.

In [386]:
df['Do_Not_Contact'] = df['Do_Not_Contact'].str.replace('Yes','Y').str.replace('No','N')
df

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1002       Abed       Nadir  123-643-9775               N   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345               Y   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467               Y   
6         1007       Jeff      Winger                             N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN                             Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee                             Y   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                           N/a   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire     None  
1               Y   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4               N      123 Dragons Road           None     None  
5               Y      768 City Parkway           None     None  
6               N     1209 South Street           None     None  
7               N         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9               N      25th Main Street       New York     None  
10              N        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13              N      343 City Parkway           None     None  
14              N         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16              N      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18              Y                   N/a           None     None  
19              N     910 Tatooine Road       Tatooine     None

## 7. Removing 'NaN' and 'N/a' (Null) values within the dataframe.

- We want to remove 'NaN' and N/a valus within the dataframe and we can try the replace() function to replace them with blank strings:

In [387]:
df.replace('NaN').replace('N/a','').replace('NaN','')

C:\Users\Alicja\AppData\Local\Temp\ipykernel_20100\1196532850.py:1: FutureWarning: DataFrame.replace without 'value' and with non-dict-like 'to_replace' is deprecated and will raise in a future version. Explicitly specify the new values instead.
  df.replace('NaN').replace('N/a','').replace('NaN','')


CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1002       Abed       Nadir  123-643-9775               N   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345               Y   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467               Y   
6         1007       Jeff      Winger                             N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf         NaN                             Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee                             Y   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                                 
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire     None  
1               Y   93 West Main Street           None     None  
2             NaN    298 Drugs Driveway           None     None  
3               Y      980 Paper Avenue   Pennsylvania    18503  
4               N      123 Dragons Road           None     None  
5               Y      768 City Parkway           None     None  
6               N     1209 South Street           None     None  
7               N         98 Clue Drive           None     None  
8             NaN      123 Middle Earth           None     None  
9               N      25th Main Street       New York     None  
10              N        612 Shire Lane          Shire     None  
11            NaN  2394 Hogwarts Avenue           None     None  
12              N      2039 Main Street           None     None  
13              N      343 City Parkway           None     None  
14              N         214 HR Avenue           None     None  
15              N  2395 Hogwarts Avenue           None     None  
16              N      121 Paper Avenue   Pennsylvania     None  
17            NaN       3498 Super Lane           None     None  
18              Y                                 None     None  
19              N     910 Tatooine Road       Tatooine     None

- That output an error.
- We encountered an error with `df.replace('NaN', '').replace('N/a', '')` because it didn’t handle **'pd.NA'** or **'None'**, **pandas’ and Python’s native missing value indicators** present in our DataFrame, but `df.fillna('')` fixed it by **replacing them with empty strings**, ensuring all missing data was cleaned **alongside literal strings** using **both methods**.
- `fillna` is needed because `replace({'NaN': '', 'N/a': ''})` only handles literal strings **'NaN'** and **'N/a'**, while `fillna('')` targets **pandas' native missing values (pd.NA or None)** to ensure all missing data is replaced with an empty string.
- We then combine the two to target both string literals containing **'NaN'** and **'N/a'** as well as pandas' native missing values and thus can output cleaned data.

In [388]:
df.replace({'NaN': '', 'N/a': ''}).fillna('')

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1002       Abed       Nadir  123-643-9775               N   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345               Y   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467               Y   
6         1007       Jeff      Winger                             N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf                                         Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee                             Y   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                                 
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire           
1               Y   93 West Main Street                          
2                    298 Drugs Driveway                          
3               Y      980 Paper Avenue   Pennsylvania    18503  
4               N      123 Dragons Road                          
5               Y      768 City Parkway                          
6               N     1209 South Street                          
7               N         98 Clue Drive                          
8                      123 Middle Earth                          
9               N      25th Main Street       New York           
10              N        612 Shire Lane          Shire           
11                 2394 Hogwarts Avenue                          
12              N      2039 Main Street                          
13              N      343 City Parkway                          
14              N         214 HR Avenue                          
15              N  2395 Hogwarts Avenue                          
16              N      121 Paper Avenue   Pennsylvania           
17                      3498 Super Lane                          
18              Y                                                
19              N     910 Tatooine Road       Tatooine

- That worked, we will now assign it to the dataframe:

In [389]:
df = df.replace({'NaN': '', 'N/a': ''}).fillna('')
df

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
1         1002       Abed       Nadir  123-643-9775               N   
2         1003     Walter       White                             N   
3         1004     Dwight     Schrute  123-543-2345               Y   
4         1005        Jon        Snow  876-678-3469               Y   
5         1006        Ron     Swanson  304-762-2467               Y   
6         1007       Jeff      Winger                             N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf                                         Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee                             Y   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent                             Y   
18        1019      Creed      Braton                                 
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire           
1               Y   93 West Main Street                          
2                    298 Drugs Driveway                          
3               Y      980 Paper Avenue   Pennsylvania    18503  
4               N      123 Dragons Road                          
5               Y      768 City Parkway                          
6               N     1209 South Street                          
7               N         98 Clue Drive                          
8                      123 Middle Earth                          
9               N      25th Main Street       New York           
10              N        612 Shire Lane          Shire           
11                 2394 Hogwarts Avenue                          
12              N      2039 Main Street                          
13              N      343 City Parkway                          
14              N         214 HR Avenue                          
15              N  2395 Hogwarts Avenue                          
16              N      121 Paper Avenue   Pennsylvania           
17                      3498 Super Lane                          
18              Y                                                
19              N     910 Tatooine Road       Tatooine

## 8. Removing all rows that contain values of 'Y' in the 'Do_Not_Contact' column or that contain no value in the 'Phone_Number' column.

- To iterate through the dataframe and find and remove rows that contain values of 'Y' in the 'Do_Not_Contact' column or that contain no value in the 'Phone_Number' column, we will need to perform a loop.
- The first loop will focus on removing values of 'Y' within the 'Do_Not_Contact' column.

In [390]:
for x in df.index:
   if df.loc[x, 'Do_Not_Contact']=='Y':
       df.drop(x, inplace=True)
df

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
2         1003     Walter       White                             N   
4         1005        Jon        Snow  876-678-3469               Y   
6         1007       Jeff      Winger                             N   
7         1008   Sherlock      Holmes  876-678-3469               N   
8         1009    Gandalf                                         Y   
9         1010      Peter      Parker  123-545-5421               Y   
10        1011    Samwise      Gamgee                             Y   
11        1012      Harry      Potter                             Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
17        1018      Clark        Kent                             Y   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire           
2                    298 Drugs Driveway                          
4               N      123 Dragons Road                          
6               N     1209 South Street                          
7               N         98 Clue Drive                          
8                      123 Middle Earth                          
9               N      25th Main Street       New York           
10              N        612 Shire Lane          Shire           
11                 2394 Hogwarts Avenue                          
12              N      2039 Main Street                          
13              N      343 City Parkway                          
14              N         214 HR Avenue                          
15              N  2395 Hogwarts Avenue                          
16              N      121 Paper Avenue   Pennsylvania           
17                      3498 Super Lane                          
19              N     910 Tatooine Road       Tatooine

- We can use the same loop as above, modified to remove the blank values for the 'Phone_Number' column.

In [391]:
for x in df.index:
   if df.loc[x, 'Phone_Number']=='':
       df.drop(x, inplace=True)
df

# We could have also used the following to drop null values:
# df = df.dropna(subset='Phone_Number'), inplace=True

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0         1001      Frodo     Baggins  123-545-5421               Y   
4         1005        Jon        Snow  876-678-3469               Y   
7         1008   Sherlock      Holmes  876-678-3469               N   
9         1010      Peter      Parker  123-545-5421               Y   
12        1013        Don      Draper  123-543-2345               Y   
13        1014     Leslie       Knope  876-678-3469               Y   
14        1015       Toby  Flenderson  304-762-2467               N   
15        1016        Ron     Weasley  123-545-5421               N   
16        1017   Michael        Scott  123-643-9775               Y   
19        1020     Anakin   Skywalker  876-678-3469               Y   

   Do_Not_Contact        Street_Address          State Zip_Code  
0               N        123 Shire Lane          Shire           
4               N      123 Dragons Road                          
7               N         98 Clue Drive                          
9               N      25th Main Street       New York           
12              N      2039 Main Street                          
13              N      343 City Parkway                          
14              N         214 HR Avenue                          
15              N  2395 Hogwarts Avenue                          
16              N      121 Paper Avenue   Pennsylvania           
19              N     910 Tatooine Road       Tatooine

- That worked! We now don't need to consider populating the blank values within the 'Do_Not_Contact' column to contain 'N', as luckily, all of the rows with blank values for 'Phone_Number' also happened to be the rows with blank values for the 'Do_Not_Contact' column.

## 9. We can reset the index and drop the old one to provide an accurate count of how many people are on the list for the staff at the call centre.

- This will improve knowing how many people that the contact centre staff either have called or will have to call and will replace a useless index column, which is missing values with a useful one that is accurate.

- One way that we can do this is to set the index to a column and then reset the index:

In [392]:
df = df.set_index('CustomerID')
df

First_Name   Last_Name  Phone_Number Paying_Customer  \
CustomerID                                                        
1001            Frodo     Baggins  123-545-5421               Y   
1005              Jon        Snow  876-678-3469               Y   
1008         Sherlock      Holmes  876-678-3469               N   
1010            Peter      Parker  123-545-5421               Y   
1013              Don      Draper  123-543-2345               Y   
1014           Leslie       Knope  876-678-3469               Y   
1015             Toby  Flenderson  304-762-2467               N   
1016              Ron     Weasley  123-545-5421               N   
1017         Michael        Scott  123-643-9775               Y   
1020           Anakin   Skywalker  876-678-3469               Y   

           Do_Not_Contact        Street_Address          State Zip_Code  
CustomerID                                                               
1001                    N        123 Shire Lane          Shire           
1005                    N      123 Dragons Road                          
1008                    N         98 Clue Drive                          
1010                    N      25th Main Street       New York           
1013                    N      2039 Main Street                          
1014                    N      343 City Parkway                          
1015                    N         214 HR Avenue                          
1016                    N  2395 Hogwarts Avenue                          
1017                    N      121 Paper Avenue   Pennsylvania           
1020                    N     910 Tatooine Road       Tatooine

In [393]:
df = df.reset_index()
df

CustomerID First_Name   Last_Name  Phone_Number Paying_Customer  \
0        1001      Frodo     Baggins  123-545-5421               Y   
1        1005        Jon        Snow  876-678-3469               Y   
2        1008   Sherlock      Holmes  876-678-3469               N   
3        1010      Peter      Parker  123-545-5421               Y   
4        1013        Don      Draper  123-543-2345               Y   
5        1014     Leslie       Knope  876-678-3469               Y   
6        1015       Toby  Flenderson  304-762-2467               N   
7        1016        Ron     Weasley  123-545-5421               N   
8        1017   Michael        Scott  123-643-9775               Y   
9        1020     Anakin   Skywalker  876-678-3469               Y   

  Do_Not_Contact        Street_Address          State Zip_Code  
0              N        123 Shire Lane          Shire           
1              N      123 Dragons Road                          
2              N         98 Clue Drive                          
3              N      25th Main Street       New York           
4              N      2039 Main Street                          
5              N      343 City Parkway                          
6              N         214 HR Avenue                          
7              N  2395 Hogwarts Avenue                          
8              N      121 Paper Avenue   Pennsylvania           
9              N     910 Tatooine Road       Tatooine

## This is the final dataframe, we can now save it to our computer locally for sharing.

In [395]:
df.to_excel(r'C:\Users\Alicja\Desktop\Charles\data_analytics\da_courses_and_projects\python\courses\pandas\data_cleaning_pandas\cleaned_data\customer_call_list_clean.xlsx', index=False)

## Complete!